src/picongpu/include/plugins/particleDetector/ParticleDetectorFunctors.hpp

```diff
@@ -0,0 +1,208 @@
+                     const uint32_t numBinsTheta,
+                     const uint32_t numBinsPhi,
+                     const uint32_t numBinsEnergy,
+                     const float_X minEnergy,
+                     const float_X maxEnergy,
+                     const bool logScale,
+                     const float3_X detectorFrameVecX,
+                     const float3_X detectorFrameVecY,
+                     const float3_X detectorFrameVecZ) :
+        calorimeterCur(NULL, PMacc::math::Size_t<DIM2>()),
+        maxTheta(maxTheta),
+        maxPhi(maxPhi),
+        numBinsTheta(numBinsTheta),
+        numBinsPhi(numBinsPhi),
+        numBinsEnergy(numBinsEnergy),
+        minEnergy(minEnergy),
+        maxEnergy(maxEnergy),
+        logScale(logScale),
+        detectorFrameVecX(detectorFrameVecX),
+        detectorFrameVecY(detectorFrameVecY),
+        detectorFrameVecZ(detectorFrameVecZ)
+    {}
+
+    HINLINE void setCalorimeterCursor(const CalorimeterCur& calorimeterCur)
+    {
+        this->calorimeterCur = calorimeterCur;
+    }
+
+    template<typename ParticlesFrame>
+    DINLINE void operator()(ParticlesFrame& particlesFrame, const uint32_t linearThreadIdx)
+    {
+        const float3_X mom = particlesFrame[linearThreadIdx][momentum_];
+        const float_X mom2 = math::dot(mom, mom);
+        float3_X dirVec = mom * math::rsqrt(mom2);
+
+        /* rotate dirVec into the detector frame */
+        using namespace PMacc::algorithms::math;
+        dirVec = float3_X(dot(this->detectorFrameVecX, dirVec),
+                          dot(this->detectorFrameVecY, dirVec),
+                          dot(this->detectorFrameVecZ, dirVec));
+
+        /* convert dirVec to theta and phi */
+        const float_X theta = atan2(dirVec.x(), dirVec.y());
+        const float_X phi = asin(dirVec.z());
+
+        if(abs(theta) < this->maxTheta && abs(phi) < this->maxPhi)
+        {
+            const float2_X calorimeterPos = particleDetector::mapThetaPhiToNormedRange(
+                theta, phi, this->maxTheta, this->maxPhi);
+
+            // theta
+            int32_t thetaBin = calorimeterPos.x() * static_cast<float_X>(numBinsTheta);
+            // catch out-of-range values
+            thetaBin = thetaBin >= numBinsTheta ? numBinsTheta - 1 : thetaBin;
+            thetaBin = thetaBin < 0 ? 0 : thetaBin;
+
+            // phi
+            int32_t phiBin = calorimeterPos.y() * static_cast<float_X>(numBinsPhi);
+            // catch out-of-range values
+            phiBin = phiBin >= numBinsPhi ? numBinsPhi - 1 : phiBin;
+            phiBin = phiBin < 0 ? 0 : phiBin;
+
+            // energy
+            const float_X weighting = particlesFrame[linearThreadIdx][weighting_];
+            const float_X normedWeighting = weighting /
+                                            static_cast<float_X>(particles::TYPICAL_NUM_PARTICLES_PER_MACROPARTICLE);
+            const float_X energy = SPEED_OF_LIGHT * math::sqrt(mom2) / weighting;
```

Comment:
The restriction to massless species is just a short-term solution, since for now a species always has to have a mass attribute, see #1209. So we cannot distinguish the massive and the massless case right now, i.e. the arbitrary choice of the massless energy formula is just of pragmatic reason.